src/Functions/getScalar.cpp

```diff
@@ -44,7 +44,8 @@ class FunctionGetScalar : public IFunction
         if (arguments.size() != 1 || !isString(arguments[0].type) || !arguments[0].column || !isColumnConst(*arguments[0].column))
             throw Exception("Function " + getName() + " accepts one const string argument", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
         auto scalar_name = assert_cast<const ColumnConst &>(*arguments[0].column).getValue<String>();
-        scalar = context.getScalar(scalar_name).getByPosition(0);
+        const Context & query_context = context.hasQueryContext() ? context.getQueryContext() : context;
```

Comment:
Ok, because scalars [are copied](https://github.com/ClickHouse/ClickHouse/blob/cd372de4175ff38dbb52df54b363041aa3001275/src/Interpreters/InterpreterSelectQuery.cpp#L322) to query context if query context exists.